Sources/AsyncHTTPClient/AsyncAwait/HTTPClient+execute.swift

```diff
@@ -0,0 +1,166 @@
+        let preparedRequest = try HTTPClientRequest.Prepared(request)
+        let response = try await executeCancellable(preparedRequest, deadline: deadline, logger: logger)
+
+        if !preparedRequest.body.canBeConsumedMultipleTimes,
+           let redirectState = redirectState,
+           let redirectURL = response.headers.extractRedirectTarget(
+               status: response.status,
+               originalURL: preparedRequest.url,
+               originalScheme: preparedRequest.poolKey.scheme
+           ) {
+            return try await self.followRedirect(
+                redirectURL: redirectURL,
+                redirectState: redirectState,
+                previousRequest: preparedRequest,
+                response: response,
+                deadline: deadline,
+                logger: logger
+            )
+        }
+        return response
+    }
+
+    private func followRedirect(
+        redirectURL: URL,
+        redirectState: RedirectState,
+        previousRequest: HTTPClientRequest.Prepared,
+        response: HTTPClientResponse,
+        deadline: NIODeadline,
+        logger: Logger
+    ) async throws -> HTTPClientResponse {
+        var redirectState = redirectState
+        try redirectState.redirect(to: redirectURL.absoluteString)
+
+        return try await self.executeAndFollowRedirectsIfNeeded(
+            previousRequest.followingRedirect(to: redirectURL, status: response.status),
+            deadline: deadline,
+            logger: logger,
+            redirectState: redirectState
+        )
+    }
+
+    private func executeCancellable(
+        _ request: HTTPClientRequest.Prepared,
+        deadline: NIODeadline,
+        logger: Logger
+    ) async throws -> HTTPClientResponse {
+        let cancelHandler = TransactionCancelHandler()
+
+        return try await withTaskCancellationHandler(operation: { () async throws -> HTTPClientResponse in
+            try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<HTTPClientResponse, Swift.Error>) -> Void in
+                let transaction = Transaction(
+                    request: request,
+                    requestOptions: .init(idleReadTimeout: nil),
+                    logger: logger,
+                    connectionDeadline: deadline,
+                    preferredEventLoop: self.eventLoopGroup.next(),
+                    responseContinuation: continuation
+                )
+                
+                // `HTTPClient.Task` conflicts with Swift Concurrency Task and `Swift.Task` doesn't work
+                _Concurrency.Task {
+                    await cancelHandler.registerTransaction(transaction)
+                }
+
+                self.poolManager.executeRequest(transaction)
+            }
+        }, onCancel: {
+            // `HTTPClient.Task` conflicts with Swift Concurrency Task and `Swift.Task` doesn't work
+            _Concurrency.Task {
+                await cancelHandler.cancel()
```

Comment:
Do we need `cancel` to be awaitable? Constructing these child tasks is a bit awkward.

Reply:
We don't need it to be awaitable but we use an actor as our synchronisation primitive so all methods are automatically async/awaitable from outside. 
I see two options:
1. We could make the current `cancel()` method private (and rename it) and add a new `nonisolated cancel()` method which would internal create the `Task` and call the `private cancel()` method.
2. We could also rewrite `TransactionCancelHandler` to use a lock instead and make it a class.

WDYT?

Reply:
We can leave it as-is, I guess. I'm always a bit nervous about creating lots of tasks but we can always change it later.